### PR TITLE
UCP/CORE: Reduce UCP request size (new approach)

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Reduce UCP request size (new approach).

## Why ?

The approach helps us to save 24 bytes.
The 16 bytes will be used in #5821 for hlist element on the common memory for both `send` and `recv` part of UCP request
now the biggest structures of send part of UCP request has the same size:
```
(gdb) p sizeof(((ucp_request_t*)0)->send.flush)
$1 = 32
(gdb) p sizeof(((ucp_request_t*)0)->send.rndv_get)
$2 = 32
(gdb) p sizeof(((ucp_request_t*)0)->send.msg_proto)
$3 = 32
```

## How ?

1. Re-pack UCP request `send` structure to use the hole created by the following fields (pending_lane + lane + multi_lane_idx = 24).
2. Make `mem_type` as `uint8_t` instead of `ucs_mem_type_t` (that's is `enum` and equal to the `int` type).
3. Remove pointer to the super request from send protocols in `send` part of UCP request and move it to the common memory of `send` part of UCP request to unite it with `send_nbx` callback that is used only by user-exposed UCP requests.

### CODE REVIEW

**Comment by dmitrygx** [comment]
@yosefe @brminich @hoopoepg this approach gives a better result than #5825 PR
could you review pls?

File: src/ucp/core/ucp_request.h [hunk 1]
```diff
@@ -121,8 +121,11 @@ struct ucp_request {
             void                    *buffer;    /* Send buffer */
             ucp_datatype_t          datatype;   /* Send type */
             size_t                  length;     /* Total length, in bytes */
-            ucs_memory_type_t       mem_type;   /* Memory type */
-            ucp_send_nbx_callback_t cb;         /* Completion callback */
+            union {
+                ucp_send_nbx_callback_t cb;         /* Completion callback */
+                ucp_request_t           *super_req; /* Super request that is used
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:126]: maybe super_req should be union with user_data? seems more logically close to it

File: src/ucp/core/ucp_request.h [hunk 2]
```diff
@@ -274,10 +268,13 @@ struct ucp_request {
                 uct_completion_t  uct_comp; /* UCT completion */
             } state;
 
+            ucp_lane_index_t      am_bw_index;    /* AM BW lane index */
+            uint8_t               mem_type;       /* Memory type */
             ucp_lane_index_t      pending_lane;   /* Lane on which request was moved
                                                    * to pending state */
             ucp_lane_index_t      lane;           /* Lane on which this request is being sent */
             ucp_lane_index_t      multi_lane_idx; /* Index of the lane with multi-send */
+
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:277]: remove?

File: src/ucp/rndv/rndv.c [hunk 3]
```diff
@@ -508,7 +507,8 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),
         length = ucp_mtu - remaining;
     } else {
         chunk = ucs_align_up((size_t)(rndv_req->send.length /
-                                      rndv_req->send.rndv_get.lanes_count
+                                      ucs_popcount(rndv_req->send.
+                                                   rndv_get.lanes_map_all)
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:511]: popcount can be slow
lets keep lane_count, remove lanes_map_all, and do UCS_MASK(lane_count) when needed

File: src/ucp/core/ucp_request.h (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h:277]: fixed

File: src/ucp/core/ucp_request.h (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h:126]: done

File: src/ucp/rndv/rndv.c (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/rndv/rndv.c:511]: not sure that is will work, since `lanes_map_all` contains only lane idx selected in `ucp_rndv_req_init_get_zcopy_lane_map()`

Comment (yosefe) [reply src/ucp/rndv/rndv.c:511]: yes, that's right
so we must keep both, cannot add popcount to datapath

Comment (yosefe) [reply src/ucp/rndv/rndv.c:511]: or, maybe unite lanes_map_avail with am_bw_index?

Comment (dmitrygx) [reply src/ucp/rndv/rndv.c:513]: > or, maybe unite lanes_map_avail with am_bw_index?

yes, good idea
will do

Comment (dmitrygx) [reply src/ucp/rndv/rndv.c]: done

**Comment by dmitrygx** [comment]
bot:pipe:retest

File: src/ucp/core/ucp_request.h (hunk 1 above)
Comment (yosefe) [reply src/ucp/core/ucp_request.h:130]: can we remove it from here?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: > can we remove it from here?

yes, this is leftover

File: src/ucp/core/ucp_request.h (hunk 2 above)
Comment (yosefe) [reply src/ucp/core/ucp_request.h:280]: seems redundant space?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: Fixed

**Review by yosefe — APPROVED** [review]

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -481,13 +481,16 @@ enum ucp_am_cb_flags {
...
+    UCP_AM_SEND_FLAG_REPLY = UCS_BIT(0),             /**< Force relevant reply
+                                                          endpoint to be passed to
+                                                          the data callback on the
+                                                          receiver. */
+    UCP_AM_SEND_FLAG_EAGER = UCS_BIT(1),             /**< Force UCP to use only
+                                                          eager protocol for AM sends. */
+    UCP_AM_SEND_FLAG_RNDV  = UCS_BIT(2),             /**< Force UCP to use only
+                                                          rendezvous protocol for
+                                                          AM sends. */
+    UCP_AM_SEND_REPLY      = UCP_AM_SEND_FLAG_REPLY  /**< Backward compatibility. */
```
Comment (brminich) [inline src/ucp/api/ucp.h:493]: this is unrelated from another PR

**Review by brminich — COMMENTED** [review]
@dmitrygx, seems squashing took several other commits from unrelated PRs

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: fixed

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
